// topgg-autoposter/dist/structs/BasePoster.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BasePoster = void 0;
const events_1 = require("events");
const sdk_1 = require("@top-gg/sdk");
class BasePoster extends events_1.EventEmitter {
    constructor(token, options) {
        var _a, _b, _c;
        super();
        this.options = options;
        this.started = false;
        if (!options)
            options = {};
        this.options = {
            interval: (_a = options.interval) !== null && _a !== void 0 ? _a : 1800000,
            postOnStart: (_b = options.postOnStart) !== null && _b !== void 0 ? _b : true,
            startPosting: (_c = options.startPosting) !== null && _c !== void 0 ? _c : true
        };
        if (this.options.interval < 900000) {
            throw new Error('Posting interval must be above 900000 (15 minutes)');
        }
        this.api = new sdk_1.Api(token);
    }
    _binder(binds) {
        this.binds = binds;
        if (this.options.startPosting) {
            if (this.binds.clientReady())
                this.start();
            else
                this.binds.waitForReady(() => {
                    this.start();
                });
        }
    }
    /**
     * Start the posting
     */
    start() {
        this.started = true;
        this._setupInterval();
    }
    /**
     * Stop the posting
     */
    stop() {
        this.started = false;
        clearInterval(this.interval);
        this.interval = null;
    }
    _setupInterval() {
        if (this.options.postOnStart)
            this.post();
        this.interval = setInterval(() => {
            if (!this.binds.clientReady())
                return;
            this.post();
        }, this.options.interval);
    }
    async post() {
        this.api.postStats(await this.binds.getStats())
            .then((data) => this.emit('posted', data))
            .catch(err => console.error(err));
    }
}
exports.BasePoster = BasePoster;
